client_linux.go

```diff
@@ -55,23 +57,45 @@ func (c *client) Devices() ([]*Device, error) {
 
 // Ports implements osClient.
 func (c *client) Ports() ([]*Port, error) {
-	msgs, err := c.execute(unix.DEVLINK_CMD_PORT_GET, netlink.Dump)
+	msgs, err := c.execute(unix.DEVLINK_CMD_PORT_GET, netlink.Dump, nil)
 	if err != nil {
 		return nil, err
 	}
 
 	return parsePorts(msgs)
 }
 
+// DpipeTables implements osClient.
+func (c *client) DpipeTables(dev *Device) ([]*DpipeTable, error) {
+	if dev == nil {
+		return nil, fmt.Errorf("invalid argument")
+	}
+	encoder := netlink.NewAttributeEncoder()
+	encoder.String(unix.DEVLINK_ATTR_BUS_NAME, dev.Bus)
+	encoder.String(unix.DEVLINK_ATTR_DEV_NAME, dev.Device)
+	data, err := encoder.Encode()
+	if err != nil {
+		return nil, err
+	}
+
+	msgs, err := c.execute(unix.DEVLINK_CMD_DPIPE_TABLE_GET, netlink.Acknowledge, data)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseDpipeTables(msgs)
+}
+
 // execute executes the specified command with additional header flags. The
 // netlink.Request header flag is automatically set.
-func (c *client) execute(cmd uint8, flags netlink.HeaderFlags) ([]genetlink.Message, error) {
+func (c *client) execute(cmd uint8, flags netlink.HeaderFlags, data []byte) ([]genetlink.Message, error) {
```

Comment:
I think it probably makes sense to make the last parameter `ae *netlink.AttributeEncoder`. If not nil, the Encode method can be called to create the data. If ae is nil, no data is passed.

Reply:
Thanks for suggestion! I changed this function.